stetho/src/main/java/com/facebook/stetho/inspector/network/AbstractAsyncPrettyPrinterFactory.java

```diff
@@ -0,0 +1,85 @@
+package com.facebook.stetho.inspector.network;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import com.facebook.stetho.common.ExceptionUtil;
+import com.facebook.stetho.common.Util;
+import com.facebook.stetho.inspector.network.AsyncPrettyPrinter;
+import com.facebook.stetho.inspector.network.AsyncPrettyPrinterFactory;
+
+import javax.annotation.Nullable;
+
+/**
+ * Abstract class for pretty printer factory that asynchronously downloads schema needed for
+ * pretty printing the payload
+ */
+public abstract class AbstractAsyncPrettyPrinterFactory implements AsyncPrettyPrinterFactory {
+
+  private static ExecutorService sExecutorService = Executors.newFixedThreadPool(1);
+
+  @Override
+  public AsyncPrettyPrinter getInstance(String headerName, String headerValue) {
+    try {
+      String uri = parseHeaderValueToUri(headerName, headerValue);
+      final Future<Response> response = sExecutorService.submit(new Request(new URL(uri)));
+      return new AsyncPrettyPrinter() {
```

Comment:
This instance needs to know the type that will result, per your comments on `AsyncPrettyPrinterFactory`.  However, there are two ways you can get that:

1. Simple have the abstract factory define a new, abstract method, which gets the resource type and forward the invocation of the `AsyncPrettyPrinter` defined here to that outer class method.
2. Modify `parseHeader...` to supply more context on the matched header:

```java
MatchResult matchHeader(...) {
  return new MatchResult(
      parseUriFromValue(headerValue), 
      Page.ResourceType.XHR);
}
```

(2) is my preference because it is more general and easier to understand, all other things equal.